bindings/java/src/org/sleuthkit/datamodel/blackboardutils/CommunicationArtifactsHelper.java

```diff
@@ -615,24 +683,55 @@ public BlackboardArtifact addCalllog(CommunicationDirection direction,
 		addAttributeIfNotZero(ATTRIBUTE_TYPE.TSK_DATETIME_END, endDateTime, attributes);
 		addCommDirectionIfKnown(direction, attributes);
 
-		if (fromAddress != null) {
-			addAttributeIfNotNull(ATTRIBUTE_TYPE.TSK_PHONE_NUMBER_FROM, fromAddress.getUniqueID(), attributes);
-			addAttributeIfNotNull(ATTRIBUTE_TYPE.TSK_NAME, fromAddress.getDisplayName(), attributes);
+		// set FROM attribute and create a caller account
+		AccountFileInstance callerAccountInstance;
+		if (StringUtils.isEmpty(callerId)) {
+			// for an Outgoing call, if no caller is specified, assume self account is the caller
+			if (direction == CommunicationDirection.OUTGOING) {
+				callerAccountInstance = selfAccountInstance;
+				addAttributeIfNotNull(ATTRIBUTE_TYPE.TSK_PHONE_NUMBER_FROM, selfAccountInstance.getAccount().getTypeSpecificID(), attributes);
+			} else { // incoming call without a caller id
+				throw new IllegalArgumentException("Caller Id not provided for incoming call.");
+			}
+		} else {
+			callerAccountInstance = createAccountInstance(accountsType, callerId);
+			addAttributeIfNotNull(ATTRIBUTE_TYPE.TSK_PHONE_NUMBER_FROM, callerId, attributes);
 		}
 
-		// Create a comma separated string of recipients
-		String toAddresses = addressListToString(toAddressList);
-		addAttributeIfNotNull(ATTRIBUTE_TYPE.TSK_PHONE_NUMBER_TO, toAddresses, attributes);
+		// Create a comma separated string of callee
+		List<AccountFileInstance> recipientAccountsList = new ArrayList();
+		String calleesStr = "";
+		if ((calleeIdsList != null && !calleeIdsList.isEmpty())) {
+			calleesStr = addressListToString(calleeIdsList);
+			addAttributeIfNotNull(ATTRIBUTE_TYPE.TSK_PHONE_NUMBER_TO, calleesStr, attributes);
+
+			for (String callee : calleeIdsList) {
+				if (!StringUtils.isEmpty(callee)) {
```

Comment:
We should cycle this list as a check at the beginning for ["","",""]. Otherwise, our recipientAccountsList will be empty and that could violate one of the constraints.

Reply:
Done.